### PR TITLE
rgw/sfs: multipart v2

### PR DESCRIPTION
**Pull request by jecluis**
There are known issues with this implementation. Some of them will only be addressed when the new versioning work is merged. Others need to be addressed independently.

Up for review, as the code base won't be changing significantly; still marking as draft because we won't be merging this now.

Signed-off-by: Joao Eduardo Luis \<joao@suse.com>

### CODE REVIEW

File: src/rgw/driver/sfs/sqlite/sqlite_multipart.cc
Comment (irq0) [inline src/rgw/driver/sfs/sqlite/sqlite_multipart.cc]: Does prefix allow characters '_' and '%' ?

If yes, prefix must be escaped (see ESCAPE in https://www.sqlite.org/lang_expr.html#:~:text=5.%20The%20LIKE%2C%20GLOB%2C%20REGEXP%2C%20MATCH%2C%20and%20extract%20operators)

File: src/rgw/driver/sfs/sqlite/sqlite_multipart.cc
Comment (irq0) [inline src/rgw/driver/sfs/sqlite/sqlite_multipart.cc]: Agree ;)

File: src/rgw/driver/sfs/sqlite/sqlite_multipart.cc
Comment (irq0) [inline src/rgw/driver/sfs/sqlite/sqlite_multipart.cc:284]: Why is update_all in a transaction? This should already be transacted as it is a single sql statement.

File: src/rgw/driver/sfs/sqlite/sqlite_multipart.cc
Comment (irq0) [inline src/rgw/driver/sfs/sqlite/sqlite_multipart.cc:305]: Why transaction (see above)?

File: src/rgw/driver/sfs/sqlite/sqlite_multipart.cc
Comment (irq0) [inline src/rgw/driver/sfs/sqlite/sqlite_multipart.cc]: make static function?

Name is a bit confusing (there are 3 functions with almost the same name).

Suggest refactoring all state change update code from the mark_.. into a "transition_multipart_state" that has the next state and allowed current state as parameters (`update set .. where state IN [1,2,3]` could be useful)

File: src/rgw/driver/sfs/multipart.h
Comment (irq0) [inline src/rgw/driver/sfs/multipart.h]: Suggest the more modern enum classes (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-class)

Is this a suplicate of enum class MultipartState ?

File: src/rgw/driver/sfs/sqlite/dbconn.h
Comment (irq0) [inline src/rgw/driver/sfs/sqlite/dbconn.h:211]: Foreign key to buckets table?

File: src/rgw/driver/sfs/sqlite/dbconn.h
Comment (irq0) [inline src/rgw/driver/sfs/sqlite/dbconn.h:220]: Foreign key to users?

Would require resolving the user to the table, but a constraint would prevent deleting a user that has multiparts open. The owner display name would also be redundant

File: src/rgw/driver/sfs/sqlite/dbconn.h
Comment (irq0) [inline src/rgw/driver/sfs/sqlite/dbconn.h]: Why "marker" and not "id"?

File: src/rgw/driver/sfs/sqlite/dbconn.h
Comment (irq0) [inline src/rgw/driver/sfs/sqlite/dbconn.h]: Suggest "size" to be consistent with versioned objects

File: src/rgw/driver/sfs/sqlite/dbconn.h
Comment (irq0) [inline src/rgw/driver/sfs/sqlite/dbconn.h]: Suggest "object" for consistency with fields that are named "object_"

File: src/rgw/driver/sfs/multipart.cc
Comment (irq0) [inline src/rgw/driver/sfs/multipart.cc:242]: Suggest: for (const auto& [index, part] : part_etags) { }

File: src/rgw/driver/sfs/multipart.cc
Comment (irq0) [inline src/rgw/driver/sfs/multipart.cc:177]: This has three state transitions and is quite long. 
Strongly suggest splitting it up into steps or at least marking them with comments.
It would improve readability and make the fact that we change states clearer. 
It would also help later when we want to restart states due to crashes (aggregating and done->realize object comes to mind)

File: src/rgw/driver/sfs/multipart.cc
Comment (irq0) [inline src/rgw/driver/sfs/multipart.cc:323]: O_TRUNC to not fail on existing file?

aggregating is retryable, isn't it?

If something crashes during aggegation we can just do it again. The objpath / objfd file is only guranteed to contain useful data after we transition to done state, so overwriting should be ok, not?

File: src/rgw/driver/sfs/multipart.cc
Comment (irq0) [inline src/rgw/driver/sfs/multipart.cc]: Suggest rename(2) instead. There is no use for the original directory entry after the close above is there?

File: src/rgw/driver/sfs/multipart.cc
Comment (irq0) [inline src/rgw/driver/sfs/multipart.cc:417]: This step deserves its own function like "realize_aggregated_multipart_into_object"

**Review by irq0 — COMMENTED** [review]
Good stuff

Optional / future / nits:
- Use the sqlite orm mapper feature to not have the manual translation between DBMultipart / DBOpMultipart
- Use SQL IN instead of greater / less than to check state change conditions. Relying on states having a numerical value that is somehow comparable could be a footgun when we change states in the future. A explicit list doesn't have this problem
- The "V2" postfix on multipart will loose meaning in the mid term and confuse in the long term. Like what happend to V1? Or is this related to a V2 feature in S3?

File: src/rgw/driver/sfs/multipart.cc
Comment (jecluis) [reply src/rgw/driver/sfs/multipart.cc:242]: I don't know of a good way to check whether the element being handled is the last element, if using the construct you are suggesting. As least, with the current `it` construct, we can check whether the iterator is at a given distance from `part_etags.cend()`.

Let me know if you know of a good alternative. :)

File: src/rgw/driver/sfs/sqlite/dbconn.h
Comment (jecluis) [reply src/rgw/driver/sfs/sqlite/dbconn.h:220]: I'll leave this to a later stage, hopefully when we validate that everything with the current user management and handling is sound.

File: src/rgw/driver/sfs/sqlite/sqlite_multipart.cc
Comment (jecluis) [reply src/rgw/driver/sfs/sqlite/sqlite_multipart.cc]: suggestions on how to achieve this? I keep going between that document and `sqlite_orm`'s implementation and can't quite figure out what to do (besides knowing that I can specify a third argument, for "escape", on `sqlite_orm`'s `like()`).

File: src/rgw/driver/sfs/sqlite/sqlite_multipart.cc
Comment (jecluis) [reply src/rgw/driver/sfs/sqlite/sqlite_multipart.cc:284]: in this case, so we can leverage `storage.changes()` to ascertain whether we changed anything.

File: src/rgw/driver/sfs/sqlite/sqlite_multipart.cc
Comment (jecluis) [reply src/rgw/driver/sfs/sqlite/sqlite_multipart.cc:305]: same reason as above.

File: src/rgw/driver/sfs/multipart.cc
Comment (jecluis) [reply src/rgw/driver/sfs/multipart.cc:323]: all that _should_ be true, but isn't at the moment. Once we move to that state, there's no way to do a complete again, because the path has marked the upload as complete.

That does make sense. I will try to find a way to do this in a simple way, but I might instead ask to push it to a later path set. This feels like a large chunk of changes.

File: src/rgw/driver/sfs/multipart.cc
Comment (jecluis) [reply src/rgw/driver/sfs/multipart.cc]: right, I'll do that.

File: src/rgw/driver/sfs/multipart.cc
Comment (jecluis) [reply src/rgw/driver/sfs/multipart.cc:177]: happy to split them after merging this, as a follow-up PR. There are other things that need doing as well.

Reason for this is that it's becoming annoying to squash things into the right place to keep them as logical changes within this PR, and I'd much rather have a clean patch set than ending up with a weird history.

File: src/rgw/driver/sfs/sqlite/sqlite_multipart.cc
Comment (irq0) [reply src/rgw/driver/sfs/sqlite/sqlite_multipart.cc]: One of the [sqlite orm unit tests](https://github.com/fnc12/sqlite_orm/blob/ff7c878af872f7987b62d825284fb25d6043af10/tests/static_tests/node_tuple.cpp#LL558C24-L558C24) suggests that something like 
`like(&DBMultipart::obj_name, fmt::format("{}%", prefix)).escape("\")` (needs replace("_", "\_"), replace("%", "\%") before) could work

File: src/rgw/driver/sfs/sqlite/sqlite_multipart.cc
Comment (irq0) [reply src/rgw/driver/sfs/sqlite/sqlite_multipart.cc:284]: storage.changes() wraps https://www.sqlite.org/c3ref/changes.html
(https://github.com/fnc12/sqlite_orm/blob/ff7c878af872f7987b62d825284fb25d6043af10/dev/storage_base.h#L129). This gives the last changes on a connection. The extra transaction should not be necessary. We BTW create a new connection by calling get_storage() here.

File: src/rgw/driver/sfs/multipart.cc
Comment (irq0) [reply src/rgw/driver/sfs/multipart.cc:242]: Makes sense. Suggest adding a const though if it doesn't need to be modified. `const auto& it` perhaps.

File: src/rgw/driver/sfs/multipart.cc
Comment (irq0) [reply src/rgw/driver/sfs/multipart.cc:177]: Follow up PR sounds good!

File: src/rgw/driver/sfs/multipart.cc
Comment (jecluis) [reply src/rgw/driver/sfs/multipart.cc:242]: Out of curiosity, wouldn't that be the end result of the `auto` as is, given we're asking for a `const` iterator anyway (i.e., `cbegin()` vs `begin()`)?

Comment (irq0) [reply src/rgw/driver/sfs/multipart.cc:242]: It wound not actually work, because `it` must be modifiable by the loop :). 
Was a bit curious myself and asked godbolt how various loop constructs and auto do types.

<details>
<summary>compiler explorer c++ insights</summary>

```
#include <iostream>
#include <vector>

int main(int argc, char** argv) {
    std::vector foo = {"a", "b", "c"};

    for (auto it = foo.cbegin(); it != foo.cend(); ++it) {
        std::cout << *it;
    }

    for (const auto& it : foo) {
        std::cout << it;
    }
  
    for (auto it : foo) {
        std::cout << it;
    }
  
    for (auto& it : foo) {
        std::cout << it;
    }
}
```


```
// Type your code here, or load an example.
#include <iostream>
#include <vector>

int main(int argc, char ** argv)
{
  std::vector<const char *, std::allocator<const char *> > foo = std::vector<const char *, std::allocator<const char *> >{std::initializer_list<const char *>{"a", "b", "c"}, std::allocator<const char *>()};
  for(__gnu_cxx::__normal_iterator<const char *const *, std::vector<const char *, std::allocator<const char *> > > it = foo.cbegin(); !__gnu_cxx::operator==(it, foo.cend()); it.operator++()) {
    std::operator<<(std::cout, it.operator*());
  }
  
  {
    std::vector<const char *, std::allocator<const char *> > & __range1 = foo;
    __gnu_cxx::__normal_iterator<const char **, std::vector<const char *, std::allocator<const char *> > > __begin1 = __range1.begin();
    __gnu_cxx::__normal_iterator<const char **, std::vector<const char *, std::allocator<const char *> > > __end1 = __range1.end();
    for(; !__gnu_cxx::operator==(__begin1, __end1); __begin1.operator++()) {
      const char *const & it = __begin1.operator*();
      std::operator<<(std::cout, it);
    }
    
  }
  {
    std::vector<const char *, std::allocator<const char *> > & __range1 = foo;
    __gnu_cxx::__normal_iterator<const char **, std::vector<const char *, std::allocator<const char *> > > __begin1 = __range1.begin();
    __gnu_cxx::__normal_iterator<const char **, std::vector<const char *, std::allocator<const char *> > > __end1 = __range1.end();
    for(; !__gnu_cxx::operator==(__begin1, __end1); __begin1.operator++()) {
      const char * it = __begin1.operator*();
      std::operator<<(std::cout, it);
    }
    
  }
  {
    std::vector<const char *, std::allocator<const char *> > & __range1 = foo;
    __gnu_cxx::__normal_iterator<const char **, std::vector<const char *, std::allocator<const char *> > > __begin1 = __range1.begin();
    __gnu_cxx::__normal_iterator<const char **, std::vector<const char *, std::allocator<const char *> > > __end1 = __range1.end();
    for(; !__gnu_cxx::operator==(__begin1, __end1); __begin1.operator++()) {
      const char *& it = __begin1.operator*();
      std::operator<<(std::cout, it);
    }
    
  }
  return 0;
}
```

</details>

File: src/rgw/driver/sfs/multipart.cc
Comment (0xavi0) [inline src/rgw/driver/sfs/multipart.cc:92]: We can, maybe, skip this `get_multipart` and rely on the `insert` coming next.
If it already exists we are already caching the possible exception anyway.

File: src/rgw/driver/sfs/sqlite/dbconn.h
Comment (0xavi0) [inline src/rgw/driver/sfs/sqlite/dbconn.h]: This include here causes systems with not much memory (32Gb) to get killed when building.
the line below:
```c++
#include "sqlite_orm.h"
```
disables optimisations for `sqlite_orm` for now.

File: src/rgw/driver/sfs/bucket.cc
Comment (0xavi0) [inline src/rgw/driver/sfs/bucket.cc:320]: some commented code.

File: src/rgw/driver/sfs/sqlite/buckets/multipart_definitions.h
Comment (0xavi0) [inline src/rgw/driver/sfs/sqlite/buckets/multipart_definitions.h:41]: If you want to remove the conversions code in a future PR I've pushed [this PR](https://github.com/aquarist-labs/ceph/pull/164) that might be helpful.
It adds the `ACLOwner` blob binding (`rgw::sal::Attrs` was already in the bindings) and we could use it to add any blob with a couple of lines.

**Review by 0xavi0 — COMMENTED** [review]
I'd need more time to review, but looks good to me so far.
Just minor things and the `#include` thing that makes `gcc` get killed in my build machine.

I think it would be also good to add possible `s3tests` that pass now (if any). Although we can also do that in a follow up PR.

I also agree on splitting the `complete` method in a future PR.

**Comment by jecluis** [comment]
@irq0 @0xavi0 all checks are finally successful. Merge away when ready.

**Comment by irq0** [comment]
> @irq0 @0xavi0 all checks are finally successful. Merge away when ready.

Awesome, I'd like to have another quick look tomorrow and merge tomorrow afternoon latest to have time for some more testing till the release.

**Review by irq0 — APPROVED** [review]

**Pull request merged**
